Switch between toggle FF and hold FF in ES (for RetroArch cores only)

chrizzo-hb opened the pull request:
### Summary

I added a new global option to ES which allows to set **FAST FORWARD** to either **Hold** or **Toggle**. Based on this setting, Libretro will use *Function* + *D-pad right* either for holding FF or for toggling FF.

### Motivation

Some users complained about having to hold *Function* + *D-pad right* and wanted to be able to use that hotkey combo to toggle FF instead. It is possible to set this up via `batocera.conf` by directly passing the RetroArch settings, however, that is somewhat annoying:

```
global.retroarch.input_hold_fast_forward_btn = nul
global.retroarch.input_toggle_fast_forward_btn = "h0right"
```

Personally, I don't care for fast forward anyway, but I figured some users might enjoy being able to achieve this in an easier way. 

### Proposed solution

I introduced a new global setting `global.fast_forward` to `es_features.yml` that can be set to either `hold` or `toggle`.

Within `libretroControllers.py` I use that setting to decide whether *Function* + *D-pad right* should be bound to `hold_fast_forward` (as it was before) or to `toggle_fast_forward` instead. If no value is set, it will default to `hold_fast_forward` so that no user would be surprised by a change in behavior.

I also made sure that both `hold_fast_forward` and `toggle_fast_forward` are cleared, so that there won't be any confusing behavior with both being bound at the same time.

chrizzo-hb committed:
Switch between holding FF and toggling FF via ES (RetroArch only)

chrizzo-hb committed:
Fixed copy/paste issue

chrizzo-hb committed:
Separated cases of set to hold and not set at all

Mikhailzrick commented:
Looks good, but a few things:

in libretroControllers.py it can be a bit more efficient this way:

```
if system.isOptSet('fast_forward') and system.config['fast_forward'] == 'toggle':
    retroarchspecials["right"] = "toggle_fast_forward"
```
It's already being set as hold by default, so when the configgen is launched if the es setting is anything other than "toggle" it actually doesn't need to change it. In the near future this could even be a GUI toggle with an off/on switch.

The other thing is you have fast_forward in the global shared dictionary but you'll want to do 1 of 2 things. Either on a per emulator bases add "fast_forward" to each emulator that's compatible in it's shared dictionary. You'll see in the yml file each emulator has a shared line where you're telling it which settings to show in ES from the global shared. This allows the setting to also show under each specific system in ES. So users could change to toggle for certain systems and not others. There is a main libretro shared dictionary, and tbh adding it there is probably what I would do, at least for now. I'm fairly confident right now toggle fast forward is only libretro, as any stand alone emulators I believe are hold only, and I'm not aware of any libretro cores that aren't compatible with fast forward.

So it would look like this:
```
libretro:
  shared: [powermode, tdp, videomode, ratio, shaderset, smooth, integerscale, bezel, bezel_stretch, hud, hud_corner, bezel.tattoo, bezel.tattoo_corner, bezel.tattoo_file, bezel.resize_tattoo, ai_service_enabled, ai_target_lang, ai_service_url, ai_service_pause, fast_forward, runahead, preemptiveframes, secondinstance, video_frame_delay_auto, vrr_runloop_enable, video_threaded, rumble_gain, audio_volume, bordersmode]
```

chrizzo-hb committed:
Revert "Separated cases of set to hold and not set at all"

This reverts commit e017a9d84d7b0a23dacea2fae6fdef4261e43e12.

chrizzo-hb committed:
on/off switch instead of hold/toggle setting for toggle fast forward

chrizzo-hb committed:
Enabled toggle fast forward for all libretro cores

chrizzo-hb committed:
Fixed doc

chrizzo-hb committed:
Further simplified toggle override

chrizzo-hb commented:
Yeah, you're totally right! I simplified the whole thing!

I renamed the variable to `toggle_fast_forward` and choices are now simply `1` (on) and `0` (off). I also tried the `preset="switch"` variant to display the option as a switch in ES, but that only works for a global option. If you want to be able to override it per game/core, you need to be able to distinguish between on, off, and **auto**. That doesn't seem to work with the switch prefix since the switch does not set variables to 0 but removes the entire line completely from `batocera.conf`, so I dropped that idea and still went with boolean choices instead.

I added the option to the shared libretro settings as you suggested. However, I also kept it in the shared global settings. I figured that most users who want to be able to toggle FF instead of holding would like to set this "globally" (for all libretro cores), so I think it's reasonable to have it as a global setting **and** as a shared libretro setting, right?

Mikhailzrick commented:
> Yeah, you're totally right! I simplified the whole thing!
> 
> I renamed the variable to `toggle_fast_forward` and choices are now simply `1` (on) and `0` (off). I also tried the `preset="switch"` variant to display the option as a switch in ES, but that only works for a global option. If you want to be able to override it per game/core, you need to be able to distinguish between on, off, and **auto**. That doesn't seem to work with the switch prefix since the switch does not set variables to 0 but removes the entire line completely from `batocera.conf`, so I dropped that idea and still went with boolean choices instead.
> 
> I added the option to the shared libretro settings as you suggested. However, I also kept it in the shared global settings. I figured that most users who want to be able to toggle FF instead of holding would like to set this "globally" (for all libretro cores), so I think it's reasonable to have it as a global setting **and** as a shared libretro setting, right?

Sorry. I never meant to not include it in the global dictionary, I meant the libretro one in addition to.

also, the preset switch should be used like this: preset:”switchon” or preset:”switchoff” to set the default state.

also also, the difference between a global setting and a per system setting in the batocera.conf file is the prefix in the line entry. You can have the same setting in both, or for as many systems as you want. It’ll just be gba.xyz=“1” or psx.xyz=“1”, etc. ES will automatically apply the system prefix to the setting line entry. Hopefully that makes sense.

Mikhailzrick commented:
I just tested trying the toggle and was getting some inconsistent behavior from what I would expect from batocera. I think knulli needs to be rebased I know some stuff relating to the new toggle / slider features were changed through V40 development. For now it's probably best to implement toggles in the future.
